### ISSUE TITLE
uberdev wrapper does not work

### ISSUE THREAD

**Issue by suresttexas00** [suresttexas00/minecraft-wrapper#17]
Wrapper starts and the server starts.. everything seems fine from console.  However, no ping is present and you cannot connect to the server:

http://i.imgur.com/mJ8G1Rr.png

The error message upon trying to connect yielded something along the lines of "connection reset by peer".

A change we made in proxy has to be suspect.

**Comment by JasonBristol** [comment]
Will take a look at this, my tests on my windows box were successful, If i recall you are running on linux so I will try the new codebase there

**Comment by JasonBristol** [comment]
I am able to replicate this issue on linux only, going to run some diffs and see what is causing this

**Comment by JasonBristol** [comment]
Out of curiosity has there been any changes for 1.9.2? Looks like proxy is set for 1.9.1 pre-3

**Comment by JasonBristol** [comment]
When attempting to leverage ***proxy.py*** from 6b86b65 which appears to be the last time proxy.py was updated since my changes, I get the following:
![image](https://cloud.githubusercontent.com/assets/3958523/14232952/8559ac5c-f987-11e5-934a-73f6a70ce35c.png)

I know this is our bleeding edge branch, is it possible this existed before any of my updates yesterday?

**Comment by suresttexas00** [comment]
No... The version prior to these changes is my production version used right now by my server.

**Comment by suresttexas00** [comment]
Well, you are correct.  My server does not implement those code lines.  If this is our problem, then I am really worried... Because  this is extreme spaghetti! ... that means this function _sometimes returns a client object and other times will return a UUID!!_

**Comment by suresttexas00** [comment]
I'll check our uber version with these line un-commented to see...

**Comment by suresttexas00** [comment]
I can't get any version to work properly on my test server.  I will roll back "just" the changes I made there.  However, I did find a small problem in `__main__` where `os` is not imported and referenced later:

```
        if self.config["General"]["shell-scripts"]:
            if os.name in ("posix", "mac"):
                self.scripts = Scripts(self)
...
    def rebootWrapper(self):
        self.halt = True
        os.system(" ".join(sys.argv) + "&")
...
    except SystemExit:
        # log.error("Wrapper.py received SystemExit")
        if not wrapper.configManager.exit:
            os.system("reset")
```

I am thinking it was a bad idea to make all pep-8 changes in one commit because there are too many code lines to review for actual changes.

I will make these two corrections and try to test this during my production server restart.

**Comment by suresttexas00** [comment]
OK, I've re-added import os and rolled back to the spaghetti version of getclientbyUUID that could  return a UUID... I'll go test on my actual player's server.

**Comment by suresttexas00** [comment]
http://i.imgur.com/Fy6TWnQ.png

The wrapper ignores all connections still.  Next hour I will try the pre-pep-8 version to be sure.  we may need to roll back the pep-8 changes, manually make the various fixes again, and then make pep-8 changes in small steps.

Speaking of which; tabs versus spaces - Ben will _never never never ever_ approve anything using spaces for indents.  I have argued at length with him about this:
https://github.com/benbaptist/minecraft-wrapper/issues/190

We will need to keep tab indenting if we ever want our version merged back.

**Comment by suresttexas00** [comment]
Here is the server console.. it does nothing, even though multiple people are trying to connect... 
http://i.imgur.com/jhWrsuQ.png

**Comment by suresttexas00** [comment]
1.9.2 has the same protocol functionality as 1.9.1 pre-3.  Although the PC numbers are different, they function the same.  1.9.1pre-3 changed cleintbound JoinGame packet:
http://wiki.vg/index.php?title=Pre-release_protocol&oldid=7552

If you modify the JoinGame packet back to 1.9/1.8 format, 1.9.2 servers are still compatible with 1.9 clients (the goal of universal connect was to allow these type connections).  I stopped working on UC for now when they released 1.9.2 since it is stable and my players are using 1.9.2 now.

**Comment by suresttexas00** [comment]
I should perhaps try using the pep-8 version first, in case that version was not the problem (maybe it was a subsequent modification)?  My only reason not to is I want to limit restarts and version tests on my poor players :(

**Comment by JasonBristol** [comment]
Yah that os import error I noticed as well and was going to push in next
PR. The pep8 changes were automated with the autopep8 package without the
flag that causes breaking changes

I can't get any version to work properly on my test server. I will roll
back "just" the changes I made there. However, I did find a small problem
in __main__ where os is not imported and referenced later:

        if self.config["General"]["shell-scripts"]:
            if os.name in ("posix", "mac"):
                self.scripts = Scripts(self)
...
    def rebootWrapper(self):
        self.halt = True
        os.system(" ".join(sys.argv) + "&")
...
    except SystemExit:
        # log.error("Wrapper.py received SystemExit")
        if not wrapper.configManager.exit:
            os.system("reset")

I am thinking it was a bad idea to make all pep-8 changes in one commit
because there are too many code lines to review for actual changes.

I will make these two corrections and try to test this during my production
server restart.

—
You are receiving this because you were assigned.
Reply to this email directly or view it on GitHub
<https://github.com/suresttexas00/minecraft-wrapper/issues/17#issuecomment-205003532>

**Comment by suresttexas00** [comment]
I'm trying to track down that os problem (when it was first removed -or never added).  I noticed it during a shutdown.. but shutdowns always worked in the past... which is weird... and they still work on the production server ... which is missing the `import os`!

**Comment by suresttexas00** [comment]
It has never had an import os.. ever.  It is even missing from the master stable branch.  Odd that has never been caught or caused issues.

**Comment by suresttexas00** [comment]
ok, PEP-8 version started and is running successfully:
https://github.com/suresttexas00/minecraft-wrapper/tree/3585812c04580f063f096546970db05cd538477b

The problem is later on...

**Comment by suresttexas00** [comment]
I will test PR#14 next (Work on #10 & #6 - Closing File Handles + Error Handling)

https://github.com/suresttexas00/minecraft-wrapper/commit/bd5022b2945d072979c59aa818a116125bf8015c

**Comment by suresttexas00** [comment]
The last working version apepars to be at fix#10 
https://github.com/suresttexas00/minecraft-wrapper/commit/e06eb8db77c2331c3cb5407424bb464239da3671

THis is where it first breaks after fix #6 : 

https://github.com/suresttexas00/minecraft-wrapper/commit/0ed3b9ca623d36ca4211e3d8530ec791229f3574

**Comment by JasonBristol** [comment]
Okay that makes sense, a little more involved changes were made for error
handling, it is more than likely the logic flow regarding sockets with the
new error handling, I.e the socket is always closed in a finally block,
this may be in error and causing pre mature d/cs

**Comment by suresttexas00** [comment]
I believe these are supposed to say `except Exception `__as__` e:`?

**Comment by suresttexas00** [comment]
Both `except Exception as e:` and `except Exception, e:` are valid.  I am only rolling back the pollServer change and try that out shortly.

**Comment by suresttexas00** [comment]
If `except Exception as e:` is more correct for py3, should we use that?

**Comment by suresttexas00** [comment]
Just removing the `pollServer` try - except did not resolve the issue.

**Comment by JasonBristol** [comment]
I think we will want to have a separate PR dedicated to py3 backporting to
encapsulate the changes in case of failure

**Comment by JasonBristol** [comment]
Now that I am back at my computer I will start to dig into this, it has to be something minor that changed, I also notice no activity in the console on linux. The error I get is that the connection was force-ably closed which is a minecraft server error, so it doesn't ever look like it hits the proxy

**Comment by JasonBristol** [comment]
Do you use the development branch for your production server?

We should keep the development branch stable as to not have to revert uberdev, these rollbacks will affect my downstream fork when I go to remerge.

As we make changes to uberdev that are stable we can roll them up into development and push a PR to Ben's master branch periodically.

In regards to spaces vs tabs, Ben is 100% wrong and quite frankly doesn't need to merge anything from our fork if he doesn't want to. Our goal here is to maintain a stable bug free version that Ben seems to be either too slow or too stubborn to adopt.

**Comment by suresttexas00** [comment]
Yea,  I have manually reverted some things to the way they were prior to fix #6 and I did substitute the `Exception as e`.  The revert does not work either.  I am spent for the week.  Next week I leave for a funeral.  I cant seem to find any version that works all the way around, so my server is broken for a bit plugin-wise.

**Comment by JasonBristol** [comment]
Yah something is not right somewhere, for now I am working on a 100% proxy rewrite since this module is crippled and is severely impacting any refactoring.

I will pull down the revert into my fork and start pushing commits, by the time you get back I should have a sizable PR for uberdev

**Comment by suresttexas00** [comment]
before we rewrite, lets just gets this back to working order like before.  Any  re-write should be yet another branch.  I did not intend uberdev to be this bleeding edge.

**Comment by JasonBristol** [comment]
Is the development branch not stable? You could create a PR from development to uberdev to revert back everything, as of right now uberdev appears to have not been stable pre #10

**Comment by suresttexas00** [comment]
development is stable... 
uberdev is stable through https://github.com/suresttexas00/minecraft-wrapper/commit/0ed3b9ca623d36ca4211e3d8530ec791229f3574 .
Some of my plugins are breaking for uberdev (my fault because they illegally accessed private wrapper methods that are being changed).

My intention is that:
- development would be totally suitable for pulling into Ben's main development or master.
- uberdev would be stable but may contain stuff not immediately ready for pulling to development.

so we probably need a new branch "experimental" for bleeding edge stuff.

additionally, any major changes should be on their own patch.  Other commits should be focused to a single purpose and at frequent intervals so that a specific commit can be blamed for a specific problem without losing other perfectly fine code.

**Comment by JasonBristol** [comment]
I agree 100%, to make sure I am on the same page:

- All minor commits share a common theme (i.e, fixing file handles)
- All major changes are committed to a patch branch that gets merged into the experimental branch when ready, this allow patches to be blamed and rolled back easily.
- When the experimental branch stabilizes, it gets merged into staging (might be easier to identify vs uberdev)
- When enough is staged it gets merged into the development branch. This branch is leveraged by you and me for our production servers due to the inability to identify when Ben will merge to master
- When development is ready, we submit a PR to Ben's master branch for release

The above sounds like the best way to keep our changes compartmentalized and prevent something like this in the future.

**Comment by suresttexas00** [comment]
yes, pretty close... but I would like to have staging/uberdev clean enough to run our servers on.  development should be what we are ready to deliver to Ben to the "general" community.  

Ben's merges with master are usually _very_ infrequent.  I think he will not even do a new master until he is ready for a "release candidate".

**Comment by suresttexas00** [comment]
I have already created the experimental branch.. I will make it default shortly.  When experimental reaches a pretty bug free point, we will do PR's to uberdev.  Minor bugs can be worked out there.

**Comment by suresttexas00** [comment]
Ok,
I have deleted master (on mine) and uberdev.

Added:
- experimental (the old broken uberdev)
- staging (a copy, at this time, of development).

**Comment by JasonBristol** [comment]
Awesome sounds good

I will refork and see where experimental is right now, still baffled as to why it works in windows right now

**Comment by suresttexas00** [comment]
Yes, usually it is Windows that has the fits with things.

**Comment by suresttexas00** [comment]
Sorry, I am just finding some of your replies.. I know Ben is wrong about the tabs/spaces

**Comment by suresttexas00** [comment]
In regards to spaces vs tabs, I know Ben is wrong, but I don't wan't his "ME IS STUBBORN" to prevent us from getting this (eventually) to the community.  Ultimately, at this time. it is his product, so...

I know it will create "hanging indents" in many places, but we should probably convert the spaces back to tabs.  I'm not sure what IDE you use, but it is pretty easy to switch it over with Pycharm (which I am using).

**Comment by JasonBristol** [comment]
Yah this won't be an issue for me personally, I use sublime text so its just a matter of a hotkey, and you are right, it is really up to him, I just don't want to sacrifice quality if we can prevent it

**Comment by JasonBristol** [comment]
This has been fixed via #18

**Comment by suresttexas00** [comment]
Nice!  I wont have time to test this (or be prepared in case something goes wrong) until next week.  I have to leave for a funeral in another state.  Looking forward to this when I return.
